### DISCUSSION TITLE
nwb file larger than the size of all its component data

### DISCUSSION THREAD

**Discussion by yzhaoinuw** [dandi/helpdesk#96] [Q&A]
For a experiment session, I have the image data in tif files and the segmentation mask data in .mat, plus some meatadata in json. Typically all these data together don't exceed 300 MB on disk. However, after successfully writing them (and no other data) into an nwb file, the resulting nwb file is around 350 MB or much more on disk, which is significantly larger than all its component data combined. This happens to me even after I apply compression when writing the nwb using `DataPipe`. Is this expected behavior?

**Comment by CodyCBakerPhD** [comment]
Have you shared this data on DANDI yet? If so I'd be happy to take a direct look at the NWB file to make suggestions about chunking and compression options that can make a big difference to optimizing file size at the end of the day.

Otherwise, I'd need you to report several key values

For imaging

1a) What is the shape of the photon series in question?
1b) What is the data type?
1c) What type of compression and what options?
1d) What chunk shape did you use?

For the ROI responses

2a) What is the shape of the ROI response series in question?
2b) What is the data type used for the flourescence? Is it DF/F?
2c) What type of mask was used? Image/pixel/voxel?
2d) What type of compression and what options?
2e) What chunk shape did you use?

If you have a significant number of ROIs, the metadata tables can also add up to a non-trivial size if compression is not used for the columns

**Reply by yzhaoinuw** [reply]
Hi @CodyCBakerPhD, thank you. I just uploaded my dataset to Dandi development server here: [https://gui-staging.dandiarchive.org/dandiset/208295/draft](https://gui-staging.dandiarchive.org/dandiset/208295/draft). It would be wonderful if you could give me some feedback.

On a higher level, for imaging I used `uint16` for `TwoPhotonSeries`. More details on shape and compression options can be found in the interactions I had with you and @bendichter in #92 and #93. 

For ROI, the shapes match those of the imaging data, except that they don't have a trivial first dimension as the imaging data, which has to conform with `TwoPhotonSeries` requirement. I used image mask and logical array (after be inspired by the comments from you and Ben) as the data type. Did not use compression on this one because `DataPipe` seemed not to work on logical arrays.

The size difference became less significant after I used logical array for image masks, but for some data the problem remains.

**Reply by CodyCBakerPhD** [reply]
Using `sub-BPN-M4_ses-20210524-m1_obj-1c8nyxo_ophys.nwb` as the reference here, I can confirm @bendichter's suspicion about upcasting. In the NWB file the data type of all 3 `TwoPhotonSeries` in `acquisition` are `float64` despite the values clearly being clearly within the range of `uint16`. The chunking layout and compression both look good; in fact, the compression ratio (5.6:1) is much larger than normal because of the inflation from the data type. So I can confirm that what you have right now is 284 MB of imaging data being inflated to ~1138 MB then compressed, taking up ~203 MB of space, which is still less than the uncompressed `uint16` size but could be improved by fixing this `dtype` issue.

Now, as for why all this data together is still larger than the original source files, I can practically guarantee at this point that it's because of the way MATLAB stores sparse logical arrays in a different manner than default NWB behavior does. What you have in this file are 3 `181` x `512` x `512` masks written as raw datasets without compression or sparse storage enabled. @lawrence-mbf, this is a good question for you: I've seen sparse datasets used in `h5py`, such as [this example](https://falexwolf.me/2017/sparse-matrices-with-h5py/), but is there a way to do that in MATLAB?

Regardless, there are two less technical ways of fixing that issue: a) enable compression on the image masks, or my personally preferred method, given how sparse these are, would be to just use a voxel mask, which would be a (`number of ROIs` x `3`) array, each row being the coordinates of each non-zero element from the image mask. The voxel mask approach is essentially directly applying the sparse representation that MATLAB/HDF5 use magically under the hood, which takes up far less space than basic compression on a logical array.

I also noticed something else that's quite important; based on the previous conversation, since these are static images the first axis should always be time, but for these series it seems to be depth (`181`) This is likely due to a common transposition issue from MATLAB (@bendichter and @lawrence-mbf can explain this better than I can)

**Reply by lawrence-mbf** [reply]
On sparse datasets: MATLAB supports [sparse matrices](https://www.mathworks.com/help/matlab/sparse-matrices.html) though its interop with the low-level HDF5 API is untested on our end. Note that MATLAB sparse matrices is limited to `double` sized data. I'm not sure how the dataset is translated through `h5py`.

Regarding dimensionality. The dimensions are flipped for better support with other row-ordered tools.

**Reply by yzhaoinuw** [reply]
Thank you @CodyCBakerPhD and @lawrence-mbf. Cody, I will have to work on your suggestions on the dytpe and compressing masks later if it's not critical for the publishing purpose, because we have a deadline coming up in a day or two to publish our data. I want to publish the dataset ASAP. Based on what you saw, especially regarding your comment,

> I also noticed something else that's quite important; based on the previous conversation, since these are static images the first axis should always be time, but for these series it seems to be depth (181) This is likely due to a common transposition issue from MATLAB

is there anything regarding the data shape or else that I should definitely address before publishing?

**Reply by CodyCBakerPhD** [reply]
> is there anything regarding the data shape or else that I should definitely address before publishing?

I would fix the dimension order, for sure. That usually means adding a bunch of transposes to things on the MATLAB side so they show up in the HDF5 file in the correct order (time [1 since static] x image width x image height x depth)

Otherwise the datasets are technically not correct as is, but it's a very tricky thing to catch through validation procedures since it is just as likely to find someone with non-volumetric non-static and only 181 frames of similar imaging data.

**Reply by yzhaoinuw** [reply]
Hi @CodyCBakerPhD, I downloaded `sub-BPN-M4_ses-20210524-m1_obj-1c8nyxo_ophys.nwb` from my dandiset. After reading it, I saw that the dimension for the image data from each channel (saved as a `TwoPhotonSeries`) is [1×512×512×181], and the dimension for the mask for each channel is [512x512x181]. See my Matlab console output below. The dimension order are correct, no?. Or did I miss something?

```matlab
>> nwb = nwbRead('.\sub-BPN-M4_ses-20210524-m1_obj-1c8nyxo_ophys.nwb');
>> nwb.acquisition.get('TwoPhotonSeriesChanA')

ans = 

  TwoPhotonSeries with properties:

                   field_of_view: []
                   imaging_plane: [1×1 types.untyped.SoftLink]
                        pmt_gain: []
                  scan_line_rate: []
                          device: []
                       dimension: []
                   external_file: []
    external_file_starting_frame: []
                          format: []
              starting_time_unit: 'seconds'
             timestamps_interval: 1
                 timestamps_unit: 'seconds'
                            data: [1×512×512×181 types.untyped.DataPipe]
                        comments: 'no comments'
                         control: []
             control_description: []
                 data_continuity: []
                 data_conversion: 1
                     data_offset: 0
                 data_resolution: -1
                       data_unit: 'arbitrary unit'
                     description: 'Two-photon series for ChanA'
                   starting_time: 0
              starting_time_rate: 0
                      timestamps: []
```

After loading the data from the DataPipe, I got
```matlab
>> data = nwb.acquisition.get('TwoPhotonSeriesChanA').data.load();
>> size(data)

ans =

     1   512   512   181
```

For the mask data, I got
```matlab
>> nwb.processing.get('ophys').nwbdatainterface.get('ImageSegmentation').planesegmentation.get('PlaneSegmentationChanA').image_mask.data

ans = 

  DataStub with properties:

    filename: '.\sub-BPN-M4_ses-20210524-m1_obj-1c8nyxo_ophys.nwb'
        path: '/processing/ophys/ImageSegmentation/PlaneSegmentationChanA/image_mask'
        dims: [512 512 181]
       ndims: 3
    dataType: 'logical'
```
I tried this for all three channels, `ChanA`, `ChanB`, and `ChanC`, they all have the same dimension order.

**Reply by CodyCBakerPhD** [reply]
>  See my Matlab console output below. The dimension order are correct, no?. Or did I miss something?

I am looking at it as a raw HDF5 file, which has everything transposed, which is also how Python (`h5py`) and AFAIK any other HDF5 API would see it.

I think MATLAB natively flips everything, which gets very confusing. Again, maybe @bendichter or @lawrence-mbf can explain this in more detail. But I would have expected the MatNWB API to automatically undo that transposition to avoid cross-API inconsistency?

**Reply by lawrence-mbf** [reply]
https://github.com/NeurodataWithoutBorders/matnwb/blob/442c6048c46a4849f45886975ea2c14d67bdb2e9/%2Btypes/%2Buntyped/DataStub.m#L31-L36

It's because DataStub also flips the `dims` property to pretend that everything is fine. This is to match dimension direction when reading from data stubs.

**Comment by bendichter** [comment]
> 1b) What is the data type?

Yes, this looks like a data upcasting issue to me. Maybe the data is stored as uint16 but is being written as float or something like that
